Documentation for SAML SSO support in Server 7.6

ggray-cb opened the pull request:
This is the documentation update covering the SAML SSO feature.

- [PRD](https://docs.google.com/document/d/1dH-xJvCiiaQcBjNpZMvEWwhrPp2IwF5Fl7suTupEn5Y/edit#heading=h.1kix3wng1rsj)
- [Doc plan](https://docs.google.com/document/d/16Vu9qD-YTi_fFpRusIPF-ZyOyNVqe5B4h-lMgoaXc_A/edit)

Changes in this PR, with links to the preview site:

- [Understanding Authentication](https://preview.docs-test.couchbase.com/doc-11167-SAML/server/7.6/learn/security/authentication-overview.html): Added SAMl as an option under [Authentication Options](https://preview.docs-test.couchbase.com/doc-11167-SAML/server/7.6/learn/security/authentication-overview.html#passing-credentials)
- Added [SAML Authentication](https://preview.docs-test.couchbase.com/doc-11167-SAML/server/7.6/learn/security/authentication-domains.html#saml-authentication) to Authentication Domains page.
-  Added new page to Manage named  [Configure SAML](https://preview.docs-test.couchbase.com/doc-11167-SAML/server/7.6/manage/manage-security/configure-saml.html) which covers the configuration process.
- Added new REST API page named [Configure SAML](https://preview.docs-test.couchbase.com/doc-11167-SAML/server/7.6/rest-api/rest-configure-saml.html)

To view the preview site, you'll need the username and pasword from [this Confluence page.](https://couchbasecloud.atlassian.net/wiki/spaces/DOCS/pages/1971224949/Docs+Team+demo+site+passwords)

Please have any feedback to me by 9/22.

ggray-cb committed:
Work in progress checkin.

ggray-cb committed:
In-progress checkin

ggray-cb committed:
Checking in gitignore to block VSCode settings from being pulled into repo.

ggray-cb committed:
Draft of REST API page done.

ggray-cb committed:
Completed draft (for now)...

timofey-barmin commented:
Hey @ggray-cb
Looks like this PR contains many unrelated to SAML changes (> 1000 files changed).
Do you mind moving everything that is unrelated to a separate PR?

timofey-barmin commented:
@ggray-cb 
I just realized the reason why this PR shows so many unrelated changes: I think you pushed the PR against wrong branch. It definitely should not be "release/7.2" because saml is being added in 7.6.

ggray-cb commented:
Sorry! That was my first PR using GitHub and I didn’t notice that it defaulted to 7.2 instead of the 7.6 which the branch was based on. I edited the PR and it’s now showing 13 changed files, which is correct.

From: Timofey Barmin ***@***.***>
Date: Thursday, September 21, 2023 at 4:31 PM
To: couchbase/docs-server ***@***.***>
Cc: Gary Gray ***@***.***>, Mention ***@***.***>
Subject: Re: [couchbase/docs-server] Documentation for SAML SSO support in Server 7.6 (PR #3345)
@ggray-cbI just realized the reason why this PR shows so many unrelated changes: I think you pushed the PR against wrong branch. It definitely should not be "release/7.2" because saml is being added in 7.6.—Reply to this email directly, view it on GitHub, or unsubscribe.You ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌

[Image removed by sender.]<https://usb.report.cybergraph.mimecast.com/alert-details/?dep=FBs7PpFDp4BNECgL1DDt0A%3D%3DhTBzjIPsYqWcU9echglqGY61BrkeP6LYqUnMVKQ5u9YRcPejOuk0foURPdKQCVzEZPwtVJmgX0E%2FtZGFNC9GTr6Cr8lI00tnASZEiNXfm%2FNPudAEoh%2FVAEFU%2FurJxPHSdAY2ClU780T1VarcBvNmOp1CcGRbklZrJA%2Bmm2iEAIZimA1To0Oau%2F5ia90mo6Q1aCm1bwNzMlaZ%2BTcdbtU85cnXQANp7UG9F8N99U1h8%2Bix1JgDw2r4mOweCe%2BMTSVlJnhUOrg7JjNr0OzgL%2BhMUp0%2BYHhxf4W02UTFG8oNXKM%2BMZnyGI4s%2BPf8HvwnooMsZdj8IwfUEnWpLd%2FSrUIkvMRKa2w%2FoflgsQwf5aAPG1v7MkaXOZNLVwFuC2csnxx0oHft6X%2BlPdVIJrt6jxwK0yGV5I%2B7urTwvF28vMHUixYtwsoARheSVILPOZabAt8swiWMx09oL3CsaNDD5Seq5olS5F%2B%2BIJuha11yB5Vkwwvh89VqCkgBfrWn0ERZTnyce77VLmv%2Bn01QcuyRKtGEHZSMwmCrixQ0v9YyP7XCnfg%2BVNRDW1qdU9wHkfEnWK9mVx3iOMCamKNiH9biHIaR%2B7j16iDm4j%2BAA2Vn%2Fg2%2B4cmLZS71%2BhavD%2BrC1ngoD%2Bep9CeRDspIXIdWSvsxltDF4C4k4a33WBNvjCSV4S9Ln78I7cgyD4SFyQkbxc0mZTYvLVwNHmy3NK2FBaiqN8SaVC9nccBgkX0rLcGAZIybfAx1WIzmD%2F3Q%2F5%2FLQEHbaScn>

@ggray-cb<https://github.com/ggray-cb>
I just realized the reason why this PR shows so many unrelated changes: I think you pushed the PR against wrong branch. It definitely should not be "release/7.2" because saml is being added in 7.6.

—
Reply to this email directly, view it on GitHub<https://github.com/couchbase/docs-server/pull/3345#issuecomment-1730256564>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/BA3J6NCCPO2SY2QJC6YHL63X3SP2TANCNFSM6AAAAAA42GZ3XY>.
You are receiving this because you were mentioned.Message ID: ***@***.***>

Disclaimer

The information contained in this communication from the sender is confidential. It is intended solely for use by the recipient and others authorized to receive it. If you are not the recipient, you are hereby notified that any disclosure, copying, distribution or taking action in relation of the contents of this information is strictly prohibited and may be unlawful.

This email has been scanned for viruses and malware, and may have been automatically archived by Mimecast, a leader in email security and cyber resilience. Mimecast integrates email defenses with brand protection, security awareness training, web security, compliance and other essential capabilities. Mimecast helps protect large and small organizations from malicious activity, human error and technology failure; and to lead the movement toward building a more resilient world. To find out more, visit our website.

timofey-barmin commented on modules/learn/pages/security/authentication-domains.adoc:
```diff
@@ -167,3 +170,25 @@ Note that use of the PAM Linux password-module requires all cluster-nodes to be
+
+Enabling SAML authentication does not prevent users who have locally defined accounts from logging into Couchbase Server directly with their username and password. In addition, you can enable both SAML authentication and LDAP authentication. However, users may become confused if you do not synchronize user names and passwords between the LDAP server and the IdP. 
+
+=== Mapping of Groups and Roles With SAML Authentication
```
This paragraph seems a bit confusing to me. I think we should describe clearly how user that is authenticated via SAML can get Couchbase roles.
The paragraph can be named something like "Assigning roles to SAML users".
The main point is that after successful authentication SAML user gets a _union_ of the following roles:
- roles assigned to the corresponding Couchbase external user
- roles assigned to Couchbase groups that are specified in that user's SAML attribute
- roles that are specified in that user's SAML attribute directly

We also should explicitly say that creation of Couchbase external user for each SAML user is optional. Basically this allows admins to manage users in IdP only, without any need to change anything in Couchbase on user creation/deletion.

ggray-cb replied:
I think the main point I wanted to make in the paragraph was that enabling SAML authentication doesn't disable locally-defined users logging in via the usual username and password. I revised the paragraph and moved it to the end of the next section. I also added your suggested clarification of how the user's roles are determined, and statement that admins can manage users in the IdP.

timofey-barmin replied:
My comment was about the "Mapping of Groups and Roles With SAML Authentication" paragraph (not about 'Enabling SAML authentication does not prevent users'). Github doesn't let me highlight the paragraph when I am adding a comment so I am attaching the comment to the very beginning of the sentence. Now I realize that it can look like the comment is for the paragraph above. Sorry for confusion.

ggray-cb replied:
I believe I've revised this in line with your comments.

timofey-barmin commented on modules/rest-api/examples/config-saml.sh:
```diff
@@ -0,0 +1,6 @@
+     http://localhost:8091/settings/saml | jq -S
+# end::get-saml[]
+
+curl -s -u Administrator:password http://localhost:8091/settings/saml | jq
```
Maybe I don't understand it but it seems like there are two identical examples in this file. Is that intentional?

ggray-cb replied:
This entire file is designed to be part of a test harness for the documentation examples. It's also something I can run manually to get updated output that I incorporate into the documentation by hand. So geting the SAML settings multiple times will show the changes made by other commands. The documentation pulls out pieces of this file (marked with the comments containing tag:: strings).

timofey-barmin commented on modules/rest-api/pages/rest-configure-saml.adoc:
```diff
@@ -0,0 +1,205 @@
+|The name of the SAML attribute from which Couchbase extracts the user's groups. Defaults to an empty string which has Couchbase Server use the list of groups defined for users in their Couchbase Server account. Only has an effect if groupsAttribute has a value.
+
+|groupsAttributeSep
+|The character the IdP uses to separate group names if it sends a list of groups the user. Defaults to `,`. Only has an effect if groupsAttribute has a value.
```
"The character" -> "list of characters"
Each character in that string will be used as a separator. For example, groupsAttributeSep = ";,", so it will split the following string "group1,group2;group3.group4" into the list of 3 groups: group1, group2, group3.group4

"the IdP uses" -> "couchbase-server uses"
IdP sends the attribute as is, while Couchbase server splits this string into the list of group names. 

"Defaults to `,`" -> "Defaults to ` ,`"
which means space and comma and be used as separators.

ggray-cb replied:
Two questions: When you have multiple separator characters in the list (like thed fault " ,") can you use both of them in the attribute and have them wprk? For example "group1,group2 group3" or is the string in the attribute only allowed to use one of the separators from the set?

Second question, what if you have multiple separators in a row in the string? For example, with the default value, is there an issue processing the attribute string "group1, group2, group3"? Or does Couchbase ignore what's basically a null value between the "," and " "?

timofey-barmin replied:
> When you have multiple separator characters in the list (like thed fault " ,") can you use both of them in the attribute and have them wprk? For example "group1,group2 group3" or is the string in the attribute only allowed to use one of the separators from the set?

Yes, both separators will be used (I think this is covered by my example). "group1,group2 group3" will be be split into three groups if separator is " ,".

> Second question, what if you have multiple separators in a row in the string? For example, with the default value, is there an issue processing the attribute string "group1, group2, group3"? Or does Couchbase ignore what's basically a null value between the "," and " "?

Null values will be ignored. So "  group1,    group2,    group3 " will give you 3 groups: "group1", "group2", "group3"

ggray-cb replied:
OK, added more detail to groupsAttributeSep and rolesAttributeSep

timofey-barmin commented on modules/rest-api/pages/rest-configure-saml.adoc:
```diff
@@ -0,0 +1,205 @@
+|The number of milliseconds Couchbase Server waits for a successful connection to the IdP when retrieving metadata. It reports an error if this periods elapses before the IdP responds. Defaults to 5000 (5 seconds).
+
+|idpMetadataOrigin
+|The HTTP protocol to use when retrieving metadata. This setting determines whether Couchbase Server uses a secure connection or not. Can be either `http` or the default `https`.
```
idpMetadataOrigin
It determines how couchbase-server gets IdP metadata.
Three values:
http: Couchbase-server downloads IdP's metadata using specified by idpMetadataURL URL, and updates it periodically (interval is set by idpMetadataRefreshIntervalS)
http_one_time: Couchbase-server downloads IdP's metadata using specified by idpMetadataURL URL, but does it one time (when SAML settings are being saved) and never updates it after
upload: Don't call any URL but user provides IdP metadata as a string. Typically used when Couchbase-server doesn't have direct access to IdP.

timofey-barmin commented on modules/rest-api/pages/rest-configure-saml.adoc:
```diff
@@ -0,0 +1,205 @@
+|The number of seconds Couchbase Server waits before refreshing the IdP's metadata. Only has an effect if you have enabled metadata refresh. Defaults to 3600 (1 hour).
+
+|idpMetadataTLSCAs
+|The certificate or certificate chain to use to validate the IdP. Must be in _Privacy Enhanced Mail_ (PEM) format. Required if idpMetadataTLSVerifyPeer is `true`.
```
"...to validate the IdP"
I think it makes sense to mention that it is used "when retrieving metadata".

Can you also add that it is only used when https is used in idpMetadataURL and idpMetadataTLSVerifyPeer is true?

ggray-cb replied:
Revised based on feedback.

timofey-barmin commented on modules/rest-api/pages/rest-configure-saml.adoc:
```diff
@@ -0,0 +1,205 @@
+|The certificate or certificate chain to use to validate the IdP. Must be in _Privacy Enhanced Mail_ (PEM) format. Required if idpMetadataTLSVerifyPeer is `true`.
+
+|idpMetadataTLSExtraOpts
+|Extra TLS options that Couchbase Server applies when connecting to the IdP. Only use this parameter if asked to by technical support.
```
> Extra TLS options that Couchbase Server applies when connecting to the IdP.

... to retrieve metadata

timofey-barmin commented on modules/rest-api/pages/rest-configure-saml.adoc:
```diff
@@ -0,0 +1,205 @@
+|Extra TLS options that Couchbase Server applies when connecting to the IdP. Only use this parameter if asked to by technical support.
+
+|idpMetadataTLSSNI
+|Sets a custom _Server Name Indication_ (SNI) for the connection from Couchbase Server to  the IdP. 
```
Please mention that we mean connection from Couchbase Server to the IdP to retrieve metadata (not any connection)

timofey-barmin commented on modules/rest-api/pages/rest-configure-saml.adoc:
```diff
@@ -0,0 +1,205 @@
+|A Boolean value that controls whether Couchbase Server verifies the IdP's certificate when fetching metadata. If set to `true`, Couchbase Server verifies the certificate. If it cannot verify the certificate, it returns an error. 
+
+|idpMetadataURL
+|The URL from which Couchbase Server retrieves the IdP's metadata.
```
If idpMetadataOrigin is http or http_one_time

timofey-barmin commented on modules/rest-api/pages/rest-configure-saml.adoc:
```diff
@@ -0,0 +1,205 @@
+|The name of a SAML attribute containing the roles that Couchbase Server should apply to the authenticated user. Defaults to an empty string, which has Couchbase Server use its own list of roles from the user's Couchbase Server account.
+
+|rolesAttributeSep
+|The character that separates multiple role names if the IdP supplies a list of roles in the SAML attribute set by rolesAttribute. Defaults to `,`.
```
Please see my comment for groupsAttributeSep.
It is a string containing list of characters that are used as separators when Couchbase server splits this string into the list of roles. By default ` ,` (space or comma).

timofey-barmin commented on modules/rest-api/pages/rest-configure-saml.adoc:
```diff
@@ -0,0 +1,205 @@
+|A Boolean value that controls whether Couchbase Server lets users reuse authentication credentials. See xref:manage:manage-security/configure-saml.adoc#dupe-check[SP Assertion Dupe Check] for a description of this parameter's values.
+
+|spBaseURLScheme
+|The scheme for URL that the IdP constructs to send an authenticated user back to Couchbase Server. Can be either the default `http` for unencrypted connections to Couchbase Server or `https` for encrypted connections.
```
This is scheme that will be used by Couchbase server in spMetadataURL (note: this time it is service provider (SP) metadata), spConsumeURL , spLogoutURL (all three are read only)

timofey-barmin commented on modules/rest-api/pages/rest-configure-saml.adoc:
```diff
@@ -0,0 +1,205 @@
+|The scheme for URL that the IdP constructs to send an authenticated user back to Couchbase Server. Can be either the default `http` for unencrypted connections to Couchbase Server or `https` for encrypted connections.
+
+|spBaseURLType
+|Sets the hostname portion of the URL that the IdP redirects the user's browser to after authentication. Defaults to `node`. See xref:manage:manage-security/configure-saml.adoc#sp-base-url[SP Base URL Type] for a description of this parameter's values. 
```
I think it is correct in general. We can also mention that these resulting URLs can be seen in spConsumeURL, in spMetadataURL, and in spLogoutURL (all three are read only)

timofey-barmin commented on modules/rest-api/pages/rest-configure-saml.adoc:
```diff
@@ -0,0 +1,205 @@
+|The public certificate that the IdP uses to verify the authenticity of Couchbase Server's requests in PEM format. The public key you supply must correspond to the private key you supply in spKey. You must set this value to enable SAML authentication.
+
+|spChain
+|One or more certificates in PEM formatthat were used to sign the Couchbase Server's certificate. Supply this value if Couchbase Server's certificate was not directly signed by a well-known signing authority.
```
typo: formatthat

timofey-barmin commented on modules/rest-api/pages/rest-configure-saml.adoc:
```diff
@@ -0,0 +1,205 @@
+|One or more certificates in PEM formatthat were used to sign the Couchbase Server's certificate. Supply this value if Couchbase Server's certificate was not directly signed by a well-known signing authority.
+
+|spConsumeURL
+|The URL to which the IdP should send authenticated users back to Couchbase Server. This value overrides the default URL Couchbase Server automatically generates based on spBaseURLType and spBaseURLScheme.
```
It is a read only value (it doesn't override anything, it is the value that is automatically generated). Its value is generated based on spBaseURLScheme, spBaseURLType, spCustomBaseURL.
The main purpose of this value is to provide the consume URL for IdP configuration.

timofey-barmin commented on modules/rest-api/pages/rest-configure-saml.adoc:
```diff
@@ -0,0 +1,205 @@
+|The private key Couchbase Server uses to sign and encrypt SAML messages in PEM format. You must supply a private key to enable SAML authentication. This parameter is write-only. A GET request only returns a string of asterisks for this value. 
+
+|spLogoutURL
+|The URL the IdP uses to log the user out of the Web Console. By default, Couchbase Server sets this value based on spBaseURLType and spBaseURLScheme.
```
It is a read only value (it doesn't override anything, it is the value that is automatically generated). Its value is generated based on spBaseURLScheme, spBaseURLType, spCustomBaseURL.
The main purpose of this value is to provide the logout URL for IdP configuration.

timofey-barmin commented on modules/rest-api/pages/rest-configure-saml.adoc:
```diff
@@ -0,0 +1,205 @@
+|Sets how long the IdP caches Couchbase Server's metadata. This value is in https://en.wikipedia.org/wiki/ISO_8601#Durations[ISO8601 duration format^]. The default value `P1M` sets this duration to one month.
+
+|spMetadataURL
+|The URL to retrieve Couchbase Server's SAML metadata. By default, Couchbase Server sets this value based on spBaseURLType and spBaseURLScheme.
```
It is a read only value (it doesn't override anything, it is the value that is automatically generated). Its value is generated based on spBaseURLScheme, spBaseURLType, spCustomBaseURL.
The main purpose of this value is to provide the metadata URL for IdP configuration.

timofey-barmin commented on modules/rest-api/pages/rest-configure-saml.adoc:
```diff
@@ -0,0 +1,205 @@
+|The URL the IdP uses to log the user out of the Web Console. By default, Couchbase Server sets this value based on spBaseURLType and spBaseURLScheme.
+
+|spMetadataCacheDuration
+|Sets how long the IdP caches Couchbase Server's metadata. This value is in https://en.wikipedia.org/wiki/ISO_8601#Durations[ISO8601 duration format^]. The default value `P1M` sets this duration to one month.
```
We can add it sets the CacheDuration attribute of the EntityDescriptor element in SP metadata (this is the metadata of Couchbase server).

timofey-barmin commented on modules/rest-api/pages/rest-configure-saml.adoc:
```diff
@@ -0,0 +1,205 @@
+|A URL for your organization. This value does not play a role in authentication. The IdP can use this value for various purposes, such as when prompting a user to authenticate.
+
+|spSessionExpire
+|String value that controls whether the Web Console expires its session when the IdP's authorization expires. When set to the default `SessionNotOnOrAfter` value, the Web Console has the user reauthenticate when the IdP's authentication expires. If set to `false`, the Web Console does not set its session length to match the IdP's authentication expiration.
```
It is correct, we can clarify though that if it is set to SessionNotOnOrAfter, the UI session will expire with respect to the 'SessionNotOnOrAfter' attribute of AuthnStatement element in assertion.

timofey-barmin commented on modules/rest-api/pages/rest-configure-saml.adoc:
```diff
@@ -0,0 +1,205 @@
+|A Boolean value that controls whether Couchbase Server signs its SAML requests to the IdP. Defaults to `true` which enables signing the requests.
+
+|spTrustedFingerprints
+|One or more certificate fingerprints that Couchbase Server can use to authenticate messages from the IdP instead of using certificates. 
```
One or more certificate fingerprints that Couchbase Server can use to authenticate messages from the IdP instead of using certificates from IdP metadata.

timofey-barmin commented on modules/rest-api/pages/rest-configure-saml.adoc:
```diff
@@ -0,0 +1,205 @@
+|One or more certificate fingerprints that Couchbase Server can use to authenticate messages from the IdP instead of using certificates. 
+
+|spTrustedFingerprintsUsage
+|String value that controls what Couchbase Server verifies using certificate fingerprints. When set to the default `metadataInitialOnly` value, it uses fingerprints to verify just the metadata it retrieves from the IdP. If set to `everything` it uses fingerprints for all SAML message verification. Only has an effect if at least one fingerprint is set in spTrustedFingerprints.
```
There are three values:
metadataOnly - it uses fingerprints  to verify signatures the IdP metadata signature it retrieves from the IdP (everytime it retrieves it)
metadataInitialOnly - it uses fingerprints  to verify the IdP metadata signature on first retrieval only (bootstrap), but after that, it uses certificates provided by IdP in IdP metadata for signature validation.
everything - it uses fingerprints for all SAML message signature verification (completely ignores certificates it receives in IdP metadata).

timofey-barmin commented on modules/rest-api/pages/rest-configure-saml.adoc:
```diff
@@ -0,0 +1,205 @@
+|String value that controls what Couchbase Server verifies using certificate fingerprints. When set to the default `metadataInitialOnly` value, it uses fingerprints to verify just the metadata it retrieves from the IdP. If set to `everything` it uses fingerprints for all SAML message verification. Only has an effect if at least one fingerprint is set in spTrustedFingerprints.
+
+|spVerifyAssertionEnvelopSig
+|A Boolean value. When set to `true`, Couchbase Server verifies the IdP's signature on the entire SAML message envelope. Only has an effect if idpMetadataTLSCAs contains the IdP's certificate. 
```
> Only has an effect if idpMetadataTLSCAs contains the IdP's certificate. 

This part is not correct. Please remove.

timofey-barmin commented on modules/rest-api/pages/rest-configure-saml.adoc:
```diff
@@ -0,0 +1,205 @@
+|A Boolean value. When set to `true`, Couchbase Server verifies the IdP's signature on the entire SAML message envelope. Only has an effect if idpMetadataTLSCAs contains the IdP's certificate. 
+
+|spVerifyAssertionSig
+|A Boolean value. When set to the default `true` value, Couchbase Server verifies the IdP's signature on the assertion portion of SAML messages. Only has an effect if idpMetadataTLSCAs contains the IdP's certificate. 
```
Please remove "Only has an effect if idpMetadataTLSCAs contains the IdP's certificate. "

timofey-barmin commented on modules/rest-api/pages/rest-configure-saml.adoc:
```diff
@@ -0,0 +1,205 @@
+|A Boolean value. When set to the default `true` value, Couchbase Server verifies the IdP's signature on the assertion portion of SAML messages. Only has an effect if idpMetadataTLSCAs contains the IdP's certificate. 
+
+|spVerifyLogoutReqSig
+|A Boolean value. When set to `true`, Couchbase Server verifies the IdP's signature on SAML logout requests. Only has an effect if idpMetadataTLSCAs contains the IdP's certificate. 
```
Please remove "Only has an effect if idpMetadataTLSCAs contains the IdP's certificate. "

timofey-barmin commented on modules/rest-api/pages/rest-configure-saml.adoc:
```diff
@@ -0,0 +1,205 @@
+|A Boolean value. When set to `true`, Couchbase Server verifies the IdP's signature on SAML logout requests. Only has an effect if idpMetadataTLSCAs contains the IdP's certificate. 
+
+|spVerifyRecipient
+a|Controls whether Couchbase Server verifies a SAMl message is addressed to it by comparing the SAML recipient with its own recipient name. Its value is one of the following:
```
Typo: "SAMl"

timofey-barmin commented on modules/rest-api/pages/rest-configure-saml.adoc:
```diff
@@ -0,0 +1,205 @@
+|spVerifyRecipient
+a|Controls whether Couchbase Server verifies a SAMl message is addressed to it by comparing the SAML recipient with its own recipient name. Its value is one of the following:
+
+* The string `consumeURL`: the default value which has Couchbase Server verify the recipient set in the SAML message matches its own consume URL. 
```
This is the url returned by spConsumeURL

timofey-barmin reviewed (commented):
Posting some comments. Please note that I haven't reviewed modules/manage/pages/manage-security/configure-saml.adoc yet. Will do it tomorrow.

timofey-barmin commented on modules/manage/pages/manage-security/configure-saml.adoc:
```diff
@@ -0,0 +1,198 @@
+
+You can enable _Structured Authentication Markup Language_ (SAML)  authentication that allows users to log into the Couchbase Server Web Console. This authentication methods offers features such as single sign on, two-factor authentication, and centralized authentication administration.
+
+To learn moer about SAML, see xref:learn:security/authentication-domains.adoc#saml-authentication[SAML Authentication]. 
```
typo: moer

timofey-barmin commented on modules/manage/pages/manage-security/configure-saml.adoc:
```diff
@@ -0,0 +1,198 @@
+
+=== Certificates
+
+SAML authentication requires encryption. You need a key and certificate for your Couchbase cluster to use when encrypt and decrypt this data. You can use a unique certificate specifically for SAML authentication. You can also use the same certificate you use for client authentication. 
```
At this point we don't require encryption in SAML authentication.
I would focus on signatures here. Encryption is somewhat extra measure in this context, the security is mostly based on signature verification.
When SP and IdP exchange SAML messages they sign them so remove party can always verify authenticity of those messages. The sending party signs the message with its private key. Remote party verifies the signature with certificate.

timofey-barmin commented on modules/manage/pages/manage-security/configure-saml.adoc:
```diff
@@ -0,0 +1,198 @@
+
+SAML authentication requires encryption. You need a key and certificate for your Couchbase cluster to use when encrypt and decrypt this data. You can use a unique certificate specifically for SAML authentication. You can also use the same certificate you use for client authentication. 
+ 
+Most IdPs expect certificates that are signed by a _Certificate Authority_ (CA) to verify authenticity. You can use a certificate signed by an intermediate certificate that was in turn signed by a CA if you also provide the intermediate certificate. For example, suppose your organization creates a certificate for you that they sign using a certificate signed by a CA. You can use your certificate  ...
```
I would drop this paragraph. It seems unrelated to me.

timofey-barmin commented on modules/manage/pages/manage-security/configure-saml.adoc:
```diff
@@ -0,0 +1,198 @@
+ 
+Most IdPs expect certificates that are signed by a _Certificate Authority_ (CA) to verify authenticity. You can use a certificate signed by an intermediate certificate that was in turn signed by a CA if you also provide the intermediate certificate. For example, suppose your organization creates a certificate for you that they sign using a certificate signed by a CA. You can use your certificate  ...
+  
+You can supply Couchbase Server with a certificate from the IdP so it can authenticate the IdP's metadata. Your IdP provides this certificate to you. The metadata includes a certificate that Couchbase Server uses to decrypt and authenticate SAML messages. 
```
There are multiple certificates that are used for different purposes. It can be confusing so I think it is important to describe the process clearly.

When CB server validates SAML message signatures (for example when it validates signature of SAML assertion), it can use either IdP's certificates from IdP's metadata or "Trusted fingerprints" (see spTrustedFingerprintsUsage when it uses what).
If it is configured to use IdP certificates from IdP metadata, it is important to receive metadata from IdP securely.
There are three secure ways to do it:
1. Use https for IdP metadata retrieval (see idpMetadataURL) and set idpMetadataTLSVerifyPeer = true (note that in order to verify server's identity another certificate needs to be provided, see idpMetadataTLSCAs)
2. Make sure IdP signs the metadata, then set idpSignsMetadata=true, and set at least one Trusted Fingerprint to verify that metadata
3. Upload IdP metadata manually (note that https is still recommended when the metadata is uploaded to Cocuhbase-server)

timofey-barmin commented on modules/manage/pages/manage-security/configure-saml.adoc:
```diff
@@ -0,0 +1,198 @@
+After you enable SAML authentication, the top of the configuration page lists several URLs you need to configure your IdP: 
+
+* *Current SP metadata URL*: the URL to retrieve your Couchbase Server's SAML metadata. A link also appears that you can use to download the metadata. Use this link if your IdP is not able to connect to Couchbase Server directly. 
+* *Current SP consume URL*: the URL the IdP sends the user to after they have authenticated.
```
This is a SAML Assertion Consumer Service (ACS) endpoint. It receives and processes SAML assertions from the identity provider (IdP)

timofey-barmin commented on modules/manage/pages/manage-security/configure-saml.adoc:
```diff
@@ -0,0 +1,198 @@
+
+* *Current SP metadata URL*: the URL to retrieve your Couchbase Server's SAML metadata. A link also appears that you can use to download the metadata. Use this link if your IdP is not able to connect to Couchbase Server directly. 
+* *Current SP consume URL*: the URL the IdP sends the user to after they have authenticated.
+* *Current SP logout URL*: the URL to which the IdP sends the user when they log out of the Web Console.
```
Logout endpoint that receives and processes SAML logout requests and responses from the identity provider (IdP)

timofey-barmin commented on modules/manage/pages/manage-security/configure-saml.adoc:
```diff
@@ -0,0 +1,198 @@
+Load IDP metadata from::
+Choose to either provide the IdP's metadata URL or to upload a snapshot of its metadata in a file. If you provide a URL, you can have Couchbase Server refresh the metadata periodically.
+Verify remote peer::
+Select to enable verification of the IdP's certificate. If you enable this option, supply a copy of the IdP's certificate in *CA Certificates*.   
```
> Select to enable verification of the IdP's certificate.

It is used only if https is used in "IdP's metadata URL "

timofey-barmin commented on modules/manage/pages/manage-security/configure-saml.adoc:
```diff
@@ -0,0 +1,198 @@
+Verify remote peer::
+Select to enable verification of the IdP's certificate. If you enable this option, supply a copy of the IdP's certificate in *CA Certificates*.   
+Validate metadata using trusted fingerprints::
+Select to enable using fingerprints instead of a certificate to authenticate the metadata from the IdP. Select *Use trusted fingerprints for metadata bootstrap only* to limit using fingerprint verification to just the initial retrieval of the metadata. Select *Always use trusted fingerprints to validate signatures in SAML...* to use fingerprints to validate all SAML messages instead of just metad ...
```
> Select to enable using fingerprints instead of a certificate to authenticate the metadata from the IdP

"to authenticate the metadata from the IdP" -> "validate signature of the metadata from the IdP"

Please also drop "instead of a certificate", because we can use https with remote peer verification and validate metadata signature at the same time.

ggray-cb commented:
Thanks Timofey! Does this complete you review?

From: Timofey Barmin ***@***.***>
Date: Tuesday, September 26, 2023 at 5:52 PM
To: couchbase/docs-server ***@***.***>
Cc: Gary Gray ***@***.***>, Mention ***@***.***>
Subject: Re: [couchbase/docs-server] Documentation for SAML SSO support in Server 7.6 (PR #3345)
@timofey-barmin commented on this pull request.In modules/manage/pages/manage-security/configure-saml.adoc:> @@ -0,0 +1,198 @@+= Configure SAML++You can enable _Structured Authentication Markup Language_ (SAML) authentication that allows users to log into the Couchbase Server  ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌ ‌

[Image removed by sender.]<https://usb.report.cybergraph.mimecast.com/alert-details/?dep=RCs47Gd9MLQ9rQ5k1zD0pg%3D%3D4G4D%2Fqsz9JyKegFgx0nSr74a4QxrCEKySYH7OSh13QtqzFtg9l7v3sV5mTCNrzZAzg7utVisr9rG3nQTrniP1F7RflrBf4uqY2I5v11L1c4LRgJ2tyLVG0L8obx2AwFPaaivkB0GzOpv6JZY7Y1mbokoL82%2BIHnVwY1UW9fWakiDEGfhL4kmgOoEB71ZGBgUlLXFMFwnxNqNtOGcRB5LsHN82oGHg9Yr7j9o7dzWoN6qzj3g3cq%2Fl1VRWFmClGC8dUizNuYN6sHkD5uHy4nVtnIx%2BnuZEesSs%2BJL2k8jm8xHfkZ1iWOXUUGP2jXo1rOgs%2FLbOtG7YbDYJzWh1GLKdWsyFiiA1MOkAgB5nZOx8itwgbGVbcLF8zGCfRdY7yszxtlTSbGOOHshm%2BJBRLkJlAUj70fRuH7epJXbxdta6gKl%2Fa6RlILJkXp2zTdJy0aHPyLVDMxHFDz4d8MRBjxTt83c%2BrBOyOQec5Jos4s7AZ8%2BW59rlBSAB%2BPpSznMCBkAL8FKDKh%2B1LgoxKekoMQsb9DoqUhnQ8NRn8uxjEZzR4iiZs18yFOQh%2BCyI8r7d%2Fxmil2stWQ98hbqszKqYiWzPLBSNBo6QUoI4HnX00AK7G3hc%2FmeWILw2Ti0tkQd%2FqYHOXy%2BCPLv92La%2BvYg8utvdi8WWBP4X0XrUfn7jd4AJ9f8cVTUAAwb9viokTFM9Mz9TOKo4RLpKKQjFFGdRthsl504iyPalOqR3LhtjmHreD7OC4gm%2B3U9b2jBbros0qim82rxzqu%2B4lwOEKlIKToMPw%3D%3D>

@timofey-barmin commented on this pull request.

________________________________

In modules/manage/pages/manage-security/configure-saml.adoc<https://github.com/couchbase/docs-server/pull/3345#discussion_r1337675615>:

> @@ -0,0 +1,198 @@

+= Configure SAML

+

+You can enable _Structured Authentication Markup Language_ (SAML)  authentication that allows users to log into the Couchbase Server Web Console. This authentication methods offers features such as single sign on, two-factor authentication, and centralized authentication administration.

+

+To learn moer about SAML, see xref:learn:security/authentication-domains.adoc#saml-authentication[SAML Authentication].

typo: moer

________________________________

In modules/manage/pages/manage-security/configure-saml.adoc<https://github.com/couchbase/docs-server/pull/3345#discussion_r1337732496>:

> +

+Before configuring SAML authentication consider the following:

+

+=== Identity Provider

+SAML authentication relies on an _Identity Provider_ (IdP) to authenticate a user with the Couchbase Server Web Console. Couchbase Server supports the following IdPs:

+

+* Okta

+* Microsoft Azure AD

+

+NOTE: Other IdPs may work. Couchbase Server was tested with the IdPs in the previous list.

+

+Your IdP may have its own requirements for interacting with services like Couchbase Server. See your IdP's documentation to understand its requirements.

+

+=== Certificates

+

+SAML authentication requires encryption. You need a key and certificate for your Couchbase cluster to use when encrypt and decrypt this data. You can use a unique certificate specifically for SAML authentication. You can also use the same certificate you use for client authentication.

At this point we don't require encryption in SAML authentication.
I would focus on signatures here. Encryption is somewhat extra measure in this context, the security is mostly based on signature verification.
When SP and IdP exchange SAML messages they sign them so remove party can always verify authenticity of those messages. The sending party signs the message with its private key. Remote party verifies the signature with certificate.

________________________________

In modules/manage/pages/manage-security/configure-saml.adoc<https://github.com/couchbase/docs-server/pull/3345#discussion_r1337737028>:

> +

+=== Identity Provider

+SAML authentication relies on an _Identity Provider_ (IdP) to authenticate a user with the Couchbase Server Web Console. Couchbase Server supports the following IdPs:

+

+* Okta

+* Microsoft Azure AD

+

+NOTE: Other IdPs may work. Couchbase Server was tested with the IdPs in the previous list.

+

+Your IdP may have its own requirements for interacting with services like Couchbase Server. See your IdP's documentation to understand its requirements.

+

+=== Certificates

+

+SAML authentication requires encryption. You need a key and certificate for your Couchbase cluster to use when encrypt and decrypt this data. You can use a unique certificate specifically for SAML authentication. You can also use the same certificate you use for client authentication.

+

+Most IdPs expect certificates that are signed by a _Certificate Authority_ (CA) to verify authenticity. You can use a certificate signed by an intermediate certificate that was in turn signed by a CA if you also provide the intermediate certificate. For example, suppose your organization creates a certificate for you that they sign using a certificate signed by a CA. You can use your certificate with an IdP that requires CA-signed certificates by including a copy of your organization's CA-signed certificate. The IdP uses the intermediate certificate to verify the authenticity of your cluster's certificate.

I would drop this paragraph. It seems unrelated to me.

________________________________

In modules/manage/pages/manage-security/configure-saml.adoc<https://github.com/couchbase/docs-server/pull/3345#discussion_r1337759684>:

> +SAML authentication relies on an _Identity Provider_ (IdP) to authenticate a user with the Couchbase Server Web Console. Couchbase Server supports the following IdPs:

+

+* Okta

+* Microsoft Azure AD

+

+NOTE: Other IdPs may work. Couchbase Server was tested with the IdPs in the previous list.

+

+Your IdP may have its own requirements for interacting with services like Couchbase Server. See your IdP's documentation to understand its requirements.

+

+=== Certificates

+

+SAML authentication requires encryption. You need a key and certificate for your Couchbase cluster to use when encrypt and decrypt this data. You can use a unique certificate specifically for SAML authentication. You can also use the same certificate you use for client authentication.

+

+Most IdPs expect certificates that are signed by a _Certificate Authority_ (CA) to verify authenticity. You can use a certificate signed by an intermediate certificate that was in turn signed by a CA if you also provide the intermediate certificate. For example, suppose your organization creates a certificate for you that they sign using a certificate signed by a CA. You can use your certificate with an IdP that requires CA-signed certificates by including a copy of your organization's CA-signed certificate. The IdP uses the intermediate certificate to verify the authenticity of your cluster's certificate.

+

+You can supply Couchbase Server with a certificate from the IdP so it can authenticate the IdP's metadata. Your IdP provides this certificate to you. The metadata includes a certificate that Couchbase Server uses to decrypt and authenticate SAML messages.

There are multiple certificates that are used for different purposes. It can be confusing so I think it is important to describe the process clearly.

When CB server validates SAML message signatures (for example when it validates signature of SAML assertion), it can use either IdP's certificates from IdP's metadata or "Trusted fingerprints" (see spTrustedFingerprintsUsage when it uses what).
If it is configured to use IdP certificates from IdP metadata, it is important to receive metadata from IdP securely.
There are three secure ways to do it:

  1.  Use https for IdP metadata retrieval (see idpMetadataURL) and set idpMetadataTLSVerifyPeer = true (note that in order to verify server's identity another certificate needs to be provided, see idpMetadataTLSCAs)
  2.  Make sure IdP signs the metadata, then set idpSignsMetadata=true, and set at least one Trusted Fingerprint to verify that metadata
  3.  Upload IdP metadata manually (note that https is still recommended when the metadata is uploaded to Cocuhbase-server)

________________________________

In modules/manage/pages/manage-security/configure-saml.adoc<https://github.com/couchbase/docs-server/pull/3345#discussion_r1337795885>:

> +

+. In the navigation bar, click *Security*.

+. In the navigation menu, click *SAML*.

+. Turn on the *Enabled* toggle.

+. Complete the following procedures to finish configuring SAML:

+.. <<configure-couchbase-metadata>>

+.. <<configure-the-idp>>

+.. <<configure-single-sign-on>>

+.. Optionally, <<configure-advanced-settings>>.

+. Click btn:[Save] to save the SAML settings.

+. Verify that Couchbase Server saved the settings by looking for a green popup stating that the settings were saved. If you do not see this popup, look for error messages at the top of the SAML Configuration page.

+

+After you enable SAML authentication, the top of the configuration page lists several URLs you need to configure your IdP:

+

+* *Current SP metadata URL*: the URL to retrieve your Couchbase Server's SAML metadata. A link also appears that you can use to download the metadata. Use this link if your IdP is not able to connect to Couchbase Server directly.

+* *Current SP consume URL*: the URL the IdP sends the user to after they have authenticated.

This is a SAML Assertion Consumer Service (ACS) endpoint. It receives and processes SAML assertions from the identity provider (IdP)

________________________________

In modules/manage/pages/manage-security/configure-saml.adoc<https://github.com/couchbase/docs-server/pull/3345#discussion_r1337798671>:

> +. In the navigation bar, click *Security*.

+. In the navigation menu, click *SAML*.

+. Turn on the *Enabled* toggle.

+. Complete the following procedures to finish configuring SAML:

+.. <<configure-couchbase-metadata>>

+.. <<configure-the-idp>>

+.. <<configure-single-sign-on>>

+.. Optionally, <<configure-advanced-settings>>.

+. Click btn:[Save] to save the SAML settings.

+. Verify that Couchbase Server saved the settings by looking for a green popup stating that the settings were saved. If you do not see this popup, look for error messages at the top of the SAML Configuration page.

+

+After you enable SAML authentication, the top of the configuration page lists several URLs you need to configure your IdP:

+

+* *Current SP metadata URL*: the URL to retrieve your Couchbase Server's SAML metadata. A link also appears that you can use to download the metadata. Use this link if your IdP is not able to connect to Couchbase Server directly.

+* *Current SP consume URL*: the URL the IdP sends the user to after they have authenticated.

+* *Current SP logout URL*: the URL to which the IdP sends the user when they log out of the Web Console.

Logout endpoint that receives and processes SAML logout requests and responses from the identity provider (IdP)

________________________________

In modules/manage/pages/manage-security/configure-saml.adoc<https://github.com/couchbase/docs-server/pull/3345#discussion_r1337802077>:

> ++

+* *Node address* sends the user back to the self-reported address of the node the user is logging into.

+* *Alternate node address* sends the user to the node's alternate address. See xref:learn:clusters-and-availability/connectivity.adoc#alternate-addresses[Alternate Addresses].

+* *Custom URL* sends the user to the URL you enter in *Custom URL* which appears when you select this option. Select this option when users use some other URL when connecting to nodes in the cluster. For example, if you use a load balancer to redirect users to nodes, enter its URL here.

+

+Key and Certificates::

+Add the Couchbase Server's private key and certificate. Either copy and paste their content or upload them by clicking bth:[Select File]. If the certificate was not directly signed by a CA, added the intermediate certificates in *Certificate chain*.

+

+### Configure the IdP

+

+Under *Identity Provider Configuration*, enter the information about your IdP:

+

+Load IDP metadata from::

+Choose to either provide the IdP's metadata URL or to upload a snapshot of its metadata in a file. If you provide a URL, you can have Couchbase Server refresh the metadata periodically.

+Verify remote peer::

+Select to enable verification of the IdP's certificate. If you enable this option, supply a copy of the IdP's certificate in *CA Certificates*.

Select to enable verification of the IdP's certificate.

It is used only if https is used in "IdP's metadata URL "

________________________________

In modules/manage/pages/manage-security/configure-saml.adoc<https://github.com/couchbase/docs-server/pull/3345#discussion_r1337803756>:

> +* *Alternate node address* sends the user to the node's alternate address. See xref:learn:clusters-and-availability/connectivity.adoc#alternate-addresses[Alternate Addresses].

+* *Custom URL* sends the user to the URL you enter in *Custom URL* which appears when you select this option. Select this option when users use some other URL when connecting to nodes in the cluster. For example, if you use a load balancer to redirect users to nodes, enter its URL here.

+

+Key and Certificates::

+Add the Couchbase Server's private key and certificate. Either copy and paste their content or upload them by clicking bth:[Select File]. If the certificate was not directly signed by a CA, added the intermediate certificates in *Certificate chain*.

+

+### Configure the IdP

+

+Under *Identity Provider Configuration*, enter the information about your IdP:

+

+Load IDP metadata from::

+Choose to either provide the IdP's metadata URL or to upload a snapshot of its metadata in a file. If you provide a URL, you can have Couchbase Server refresh the metadata periodically.

+Verify remote peer::

+Select to enable verification of the IdP's certificate. If you enable this option, supply a copy of the IdP's certificate in *CA Certificates*.

+Validate metadata using trusted fingerprints::

+Select to enable using fingerprints instead of a certificate to authenticate the metadata from the IdP. Select *Use trusted fingerprints for metadata bootstrap only* to limit using fingerprint verification to just the initial retrieval of the metadata. Select *Always use trusted fingerprints to validate signatures in SAML...* to use fingerprints to validate all SAML messages instead of just metadata.

Select to enable using fingerprints instead of a certificate to authenticate the metadata from the IdP

"to authenticate the metadata from the IdP" -> "validate signature of the metadata from the IdP"

Please also drop "instead of a certificate", because we can use https with remote peer verification and validate metadata signature at the same time.

—
Reply to this email directly, view it on GitHub<https://github.com/couchbase/docs-server/pull/3345#pullrequestreview-1645028583>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/BA3J6NE2T56KZ5MYXIMWF7TX4NE75ANCNFSM6AAAAAA42GZ3XY>.
You are receiving this because you were mentioned.Message ID: ***@***.***>

Disclaimer

The information contained in this communication from the sender is confidential. It is intended solely for use by the recipient and others authorized to receive it. If you are not the recipient, you are hereby notified that any disclosure, copying, distribution or taking action in relation of the contents of this information is strictly prohibited and may be unlawful.

This email has been scanned for viruses and malware, and may have been automatically archived by Mimecast, a leader in email security and cyber resilience. Mimecast integrates email defenses with brand protection, security awareness training, web security, compliance and other essential capabilities. Mimecast helps protect large and small organizations from malicious activity, human error and technology failure; and to lead the movement toward building a more resilient world. To find out more, visit our website.

timofey-barmin commented:
@ggray-cb 
Yes, this completes my review.

ggray-cb committed:
edits based on Timofey's feedback.

ggray-cb committed:
More edits based on Timofey's feedback. Mainly clarifying how certificates are used.

ggray-cb committed:
Edits based on Timofey's feedback plus general cleanup.

ggray-cb committed:
Minor rewording.

ggray-cb committed:
Fixed issue with spVerifyAssertionSig description.

timofey-barmin commented on modules/manage/pages/manage-security/configure-saml.adoc:
```diff
@@ -18,16 +18,26 @@ NOTE: Other IdPs may work. Couchbase Server was tested with the IdPs in the prev
-Instead of using an IdP-provided certificate, you can use trusted fingerprints to verify the authenticity the IdP's metadata or all of the IdP's SAML messages. Using fingerprints reduces the overhead of verification. 
+If you choose to use certificates, you must make sure Couchbase Server gets the metadata from the IdP securely. You have three options to securely get the metadata:
+
+* Give Couchbase Server a secure HTTP URL to retrieve the metadata from the IdP directly. Then enable **Verify remote peer** on the SAML configuration page and supply the certificate the IdP uses to sign its HTTPS responses. This certificate is often different from the certificate it uses to sign its SAML metadata. With this configuration, Couchbase Server verifies the identity of the IdP when do ...
```
1. HTTP -> HTTPS: "Give Couchbase Server a secure __HTTPS__ URL"
2. Suggest "...and supply the certificate the IdP uses to sign its HTTPS responses" -> "...and supply the CA certificate that issued the IdP metadata server certificate", or something similar

timofey-barmin commented on modules/manage/pages/manage-security/configure-saml.adoc:
```diff
@@ -18,16 +18,26 @@ NOTE: Other IdPs may work. Couchbase Server was tested with the IdPs in the prev
+If you choose to use certificates, you must make sure Couchbase Server gets the metadata from the IdP securely. You have three options to securely get the metadata:
+
+* Give Couchbase Server a secure HTTP URL to retrieve the metadata from the IdP directly. Then enable **Verify remote peer** on the SAML configuration page and supply the certificate the IdP uses to sign its HTTPS responses. This certificate is often different from the certificate it uses to sign its SAML metadata. With this configuration, Couchbase Server verifies the identity of the IdP when do ...
+* If the IdP signs its metadata, enable **Validate data using trusted fingerprints** on the SAML configuration page. Then add a trusted fingerprint for the IdP's certificate to the **Trusted Fingerprints** box. Couchbase Server verifies the IdP's signature in the metadata using the fingerprint and therefore trusts the certificate in the metadata.
```
"Validate data using trusted fingerprints" -> "Validate metadata using trusted fingerprints"

timofey-barmin reviewed (commented):
Thanks! Only few minor comments and we are done

ggray-cb committed:
Addressing comments in Timofey's second round of reviews.

timofey-barmin reviewed (approved):
LGTM

ggray-cb committed:
Merge branch 'release/7.6' into DOC-11167-SAML-SSO-support

ggray-cb committed:
Minor edit to appease Vale.

ggray-cb committed:
Fixed broken link.

dijarvrella commented:
@ggray-cb @bfavini @timofey-barmin sorry but can you please clarify when this will be available? I understand release/7.6 should that also correspond to the docker image 7.6?

Also stumbled upon this: https://www.couchbase.com/products/whats-new/#server however I can't find the image anywhere? Otherwise what is the guide you can share for building our own image?
